.github/workflows/build.yml

```diff
@@ -61,28 +76,54 @@ jobs:
         run: |
           python do.py setup
           python do.py init
-      - name: Build distribution
-        run: |
-          python do.py dist
-      - name: Install package on clean env
-        run: |
-          rm -rf .env
-          python do.py setup
-          python do.py install
-      - name: Run artifact generation
+      - uses: actions/download-artifact@v2
+        with:
+          name: generated-artifacts
+      - name: Display structure of downloaded files
+        run: ls -R
+      - name: Run python tests
         run: |
-          python do.py generate
-      - name: Run go lint
+          sudo python do.py test_python
+
+  go_tests:
+    needs: python_tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.head_ref }}
+      - uses: actions/download-artifact@v2
+        with:
+          name: generated-artifacts
+      - name: Display structure of downloaded files
+        run: ls -R
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.17.5"
+      - name: Setup protoc
         run: |
-          python do.py go_lint
-      - name: Run tests
+          python do.py setup_ext
+      - name: Run go tests
         run: |
-          sudo python do.py test
-      - name: Archive generated artifacts
-        uses: actions/upload-artifact@v2
+          python do.py test_go
+
+  publish_python_package:
+    needs: go_tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
         with:
-          name: generated-artifacts
-          path: art
+          ref: ${{ github.head_ref }}
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+      - name: Generate only for 3.9
```

Comment:
can't we just download artifacts and post them (since that's the one we've tested) ?